<?xml version="1.0"?>
<robot name="px100" 
  xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:arg name="moveit" default="false"/>

  <xacro:if value="$(arg moveit)">
    <link name="world"/>
    <joint name="fixed" type="fixed">
      <parent link="world"/>
      <child link="arm_base_link"/>
    </joint>
  </xacro:if>

  <link name="arm_base_link">
    <visual>
      <origin rpy="0 0 ${pi/2}" xyz="0.05 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-1-Base.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-1-Base.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 ${pi/2}" xyz="-0.0332053000 0.0008915770 0.0211913000"/>
      <mass value="0.395887" />
      <inertia ixx="0.0010650000" iyy="0.0003332000" izz="0.0012080000" ixy="-0.0000130300" ixz="0.0000018614" iyz="0.0000409200" />
    </inertial>
  </link>

  <joint name="waist" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="100" lower="${-pi}" upper="${pi}" velocity="${pi}"/>
    <origin rpy="0 0 0" xyz="0 0 0.0508"/>
    <parent link="arm_base_link"/>
    <child link="arm_shoulder_link"/>
  </joint>

  <transmission name="trans_waist">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="waist">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="waist_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_shoulder_link">
    <visual>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 -0.0022"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-2-Shoulder.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 -0.0022"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-2-Shoulder.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 ${pi/2}" xyz="0.0000111169 -0.0003605640 0.0284598000"/>
      <mass value="0.072587" />
      <inertia ixx="0.0000231000" iyy="0.0000253500" izz="0.0000144200" ixy="0.0000000003" ixz="0.0000001606" iyz="-0.0000000206" />
    </inertial>
  </link>

  <joint name="shoulder" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="${radians(-111)}" upper="${radians(107)}" velocity="${pi}"/>
    <origin rpy="0 0 0" xyz="0 0 0.04225"/>
    <parent link="arm_shoulder_link"/>
    <child link="arm_upper_arm_link"/>
  </joint>

  <transmission name="trans_shoulder">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="shoulder">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="shoulder_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_upper_arm_link">
    <visual>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-3-UA.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-3-UA.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 ${pi/2}" xyz="0.0161976963 -0.0002929352 0.0877230000"/>
      <mass value="0.082923" />
      <inertia ixx="0.0000800600" iyy="0.0000745900" izz="0.0000368500" ixy="-0.0000002144" ixz="0.0000002982" iyz="0.0000165700" />
    </inertial>
  </link>

  <joint name="elbow" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="${radians(-92)}" upper="${radians(121)}" velocity="${pi}"/>
    <origin rpy="${pi} 0 0" xyz="0.035 0 0.1"/>
    <parent link="arm_upper_arm_link"/>
    <child link="arm_forearm_link"/>
  </joint>

  <transmission name="trans_elbow">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="elbow">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="elbow_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_forearm_link">
    <visual>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-4-Forearm.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 0 ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-4-Forearm.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 ${pi/2}" xyz="0.0773720000 -0.0003324882 0.0000000000"/>
      <mass value="0.073058" />
      <inertia ixx="0.0000533800" iyy="0.0000165300" izz="0.0000603500" ixy="-0.0000003073" ixz="0.0000000000" iyz="0.0000000000" />
    </inertial>
  </link>

  <joint name="wrist_angle" type="revolute">
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="${radians(-123)}" upper="${radians(100)}" velocity="${pi}"/>
    <origin rpy="0 0 0" xyz="0.1 0 0"/>
    <parent link="arm_forearm_link"/>
    <child link="arm_gripper_link"/>
  </joint>

  <transmission name="trans_wrist_angle">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wrist_angle">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="wrist_angle_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_gripper_link">
    <visual>
      <origin rpy="0 ${pi} ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-5-Wrist-Gripper.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 ${pi} ${pi/2}" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-5-Wrist-Gripper.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 ${pi} ${pi/2}" xyz="0.0446910000 0.0000000000 -0.0113540000"/>
      <mass value="0.069929" />
      <inertia ixx="0.0000226800" iyy="0.0000204400" izz="0.0000197400" ixy="0.0000000000" ixz="0.0000000000" iyz="0.0000008485" />
    </inertial>
  </link>

  <joint name="ee_arm" type="fixed">
    <axis xyz="1 0 0"/>
    <origin rpy="${-pi} 0 0" xyz="0.063 0 0"/>
    <parent link="arm_gripper_link"/>
    <child link="arm_ee_arm_link"/>
  </joint>

  <link name="arm_ee_arm_link">
    <inertial>
      <mass value="0.001" />
      <inertia ixx="0.001" iyy="0.001" izz="0.001" ixy="0" ixz="0" iyz="0" />
    </inertial>
  </link>

  <joint name="gripper" type="continuous">
    <axis xyz="1 0 0"/>
    <limit effort="100" velocity="${pi}"/>
    <origin rpy="0 0 0" xyz="0.0055 0 0"/>
    <parent link="arm_ee_arm_link"/>
    <child link="arm_gripper_prop_link"/>
  </joint>

  <transmission name="trans_gripper">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gripper">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="gripper_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_gripper_prop_link">
    <visual>
      <origin rpy="0 0 ${pi/2}" xyz="-0.0685 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-6-Gripper-Prop.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 0 ${pi/2}" xyz="-0.0685 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-6-Gripper-Prop.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 ${pi/2}" xyz="0.0008460000 -0.0000019330 0.0000420000"/>
      <mass value="0.00434" />
      <inertia ixx="0.0000005923" iyy="0.0000011156" izz="0.0000005743" ixy="0.0000000000" ixz="0.0000003195" iyz="-0.0000000004" />
    </inertial>
  </link>

  <joint name="gripper_bar" type="fixed">
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="arm_ee_arm_link"/>
    <child link="arm_gripper_bar_link"/>
  </joint>

  <link name="arm_gripper_bar_link">
    <visual>
      <origin rpy="0 0 ${pi/2}" xyz="-0.063 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-7-Gripper-Bar.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 0 ${pi/2}" xyz="-0.063 0 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-7-Gripper-Bar.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 ${pi/2}" xyz="0.0096870000 0.0000005660 0.0049620000"/>
      <mass value="0.034199" />
      <inertia ixx="0.0000074125" iyy="0.0000284300" izz="0.0000286000" ixy="-0.0000000008" ixz="-0.0000000006" iyz="-0.0000013889" />
    </inertial>
  </link>

  <joint name="ee_bar" type="fixed">
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.023 0 0"/>
    <parent link="arm_gripper_bar_link"/>
    <child link="arm_fingers_link"/>
  </joint>

  <link name="arm_fingers_link">
    <inertial>
      <mass value="0.001" />
      <inertia ixx="0.001" iyy="0.001" izz="0.001" ixy="0" ixz="0" iyz="0" />
    </inertial>
  </link>

  <joint name="left_finger" type="prismatic">
    <axis xyz="0 1 0"/>
    <limit effort="10" lower="0.015" upper="0.037" velocity="1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="arm_fingers_link"/>
    <child link="arm_left_finger_link"/>
  </joint>

  <transmission name="trans_left_finger">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="left_finger">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="left_finger_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_left_finger_link">
    <visual>
      <origin rpy="${pi} ${pi} 0" xyz="0 0.005 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-8-Finger.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="${pi} ${pi} 0" xyz="0 0.005 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-8-Finger.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="${pi} ${pi} ${pi/2}" xyz="0.0138160000 0.0000000000 0.0000000000"/>
      <mass value="0.016246" />
      <inertia ixx="0.0000047310" iyy="0.0000015506" izz="0.0000037467" ixy="-0.0000004560" ixz="0.0000000000" iyz="0.0000000000" />
    </inertial>
  </link>

  <joint name="right_finger" type="prismatic">
    <axis xyz="0 1 0"/>
    <limit effort="10" lower="-0.037" upper="-0.015" velocity="1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="arm_fingers_link"/>
    <child link="arm_right_finger_link"/>
  </joint>

  <transmission name="trans_right_finger">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="right_finger">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="right_finger_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <link name="arm_right_finger_link">
    <visual>
      <origin rpy="0 ${pi} 0" xyz="0 -0.005 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-8-Finger.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="dark"/>
    </visual>
    <collision>
      <origin rpy="0 ${pi} 0" xyz="0 -0.005 0"/>
      <geometry>
        <mesh filename="package://rss_description/meshes/px100/PXA-100-M-8-Finger.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 ${pi} ${pi/2}" xyz="0.0138160000 0.0000000000  0.0000000000"/>
      <mass value="0.016246" />
      <inertia ixx="0.0000047310" iyy="0.0000015506" izz="0.0000037467" ixy="0.0000004560" ixz="0.0000000000" iyz="0.0000000000" />
    </inertial>
  </link>

  <joint name="ee_gripper" type="fixed">
    <axis xyz="1 0 0"/>
    <origin rpy="0 0 0" xyz="0.027575 0 0"/>
    <parent link="arm_fingers_link"/>
    <child link="arm_ee_gripper_link"/>
  </joint>

  <link name="arm_ee_gripper_link">
    <inertial>
      <mass value="0.001" />
      <inertia ixx="0.001" iyy="0.001" izz="0.001" ixy="0" ixz="0" iyz="0" />
    </inertial>
  </link>

  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
      <legacyModeNS>true</legacyModeNS>
    </plugin>
  </gazebo>

</robot>
